---
title: "Figure3"
author: "cahaisv"
date: "21 mars 2019"
output: 
  html_document:
     code_folding: "show"
params:
  reload: FALSE
---

```{r include = TRUE}
library(data.table)
suppressPackageStartupMessages(library(reshape2))
suppressPackageStartupMessages(library(gplots))
suppressPackageStartupMessages(library(ggplot2))
library(ggrepel)
library(knitr)
library(formattable)
library(limma)
library(edgeR)
suppressPackageStartupMessages(library(Homo.sapiens))
library(circlize)
library(grid)
suppressPackageStartupMessages(library(gridExtra))
```

```{r include = FALSE}
#colors
upregulated<-"indianred"
dowregulated<-"darkgreen"
palette <- colorRampPalette(c(dowregulated, "#FFFFFF", upregulated))
snv_col<-"#7570B3" #sinle nucleotide variants
amp_col<-"#D95F02" #amplification cnv+
amp_m_col<-"#B93F02" 
del_col<-"#1B9E77" #00b050 #deletions cnv-
del_m_col<-"#01796f" ## #F2AD00
mul_col<-"#333344" ## ##0B1F02 #ffc000 #multiple alterations
palette1<-c(snv_col, amp_col, del_col, mul_col)
palette2<-c(snv_col, amp_col, amp_m_col, del_col, del_m_col, mul_col)
palette3<-c("#0070c0","#ff0000","#7030a0","#c5e0b4","#e5a67b","#8faadc")
pal1 = c("darkblue", "blue","white","red","purple")
pal2 = c("darkblue", "blue","white","green","darkgreen")
palette4=c("#097C55", "#B73D00", "#535091")
palette5=c("#1B9E77", "#D95F02", "#7570B3")
```

```{r include = TRUE}
genelist<-list( 
  HM_e=c("JMJD1C","JMJD6","JMJD8","KDM1A","KDM1B","KDM2A","KDM2B","KDM3A","KDM3B","KDM4A","KDM4B","KDM4C","KDM4D","KDM4E","KDM5A","KDM5B","KDM5C","KDM5D","KDM6A","KDM6B","KDM7A","KDM8","UTY","HR","MINA"),
  HM_w=c("ASH1L","ASH2L","CARM1","DOT1L","EHMT1","EHMT2","EZH1","EZH2","MLLT10","MLLT6","NSD1","PRDM1","PRDM2","PRDM4","PRDM5","PRDM6","PRDM7","PRDM8","PRDM9","PRDM10","PRDM11","PRDM12","PRDM13","PRDM14","PRDM15","PRDM16","PRMT1","PRMT2","PRMT3","PRMT5","PRMT6","PRMT7","PRMT8","SETD1A","SETD1B","SETD2","SETD3","SETD4","SETD5","SETD6","SETD7","SETD8","SETDB1","SETDB2","SETMAR","SMYD1","SMYD2","SMYD3","SMYD4","SMYD5","SUV39H1","SUV39H2","SUV420H1","SUV420H2","AEBP2","KMT2A","KMT2B","KMT2C","KMT2D","KMT2E","WHSC1","WHSC1L1"),
  HM_r=c("CDYL","CDYL2","PHF1","PHF12","PHF19","PHF20","PHF21A","PHF21B","PHF23","CBX1","CBX3","CBX4","CBX5","CBX6","CBX7","TDRD3","SND1","MTF2","AIRE","ING1","ING2","ING3","ING4","ING5","PYGO1","PYGO2","GLYR1","MSH6","PSIP1","ZCWPW1","ZCWPW2","CCDC101","HDGFRP2","MORF4L1","MPHOSPH8","RAG2","SCML2","SFMBT1"),
  HA_e=c("HDAC1","HDAC2","HDAC3","HDAC4","HDAC5","HDAC6","HDAC7","HDAC8","HDAC9","HDAC10","HDAC11","SIRT1","SIRT2","SIRT3","SIRT4","SIRT5","SIRT6","SIRT7"),
  HA_w=c("CLOCK","CREBBP","ELP3","EP300","GTF3C4","HAT1","KAT2A","KAT2B","KAT5","KAT6A","KAT6B","KAT7","KAT8","ELP4","NCOA1","NCOA3","MSL3","KANSL1"),
  HA_r=c("ATAD2","ATAD2B","BAZ1B","BAZ1A","BAZ2A","BAZ2B","BPTF","BRD1","BRD2","BRD3","BRD4","BRD7","BRD8","BRD9","BRDT","BRPF1","BRPF3","BRWD1","BRWD3","CECR2","PBRM1","PHIP","SP100","SP110","SP140","SP140L","TAF1","TAF1L","TAF3","TRIM24","TRIM28","TRIM33","TRIM66","ZMYND11","ZMYND8","DPF3"),
  DM_w=c("DNMT1","DNMT3A","DNMT3B","DNMT3L"),
  DM_e=c("AICDA","TET1","TET2","TET3","IDH1","IDH2"),
  DM_r=c("MBD1","MBD2","MBD3","MBD4","MBD5","MECP2","UHRF1","UHRF2"),
  ChRC=c("RSF1","POLE3","CHRAC1","RBBP7","RBBP4","MTA1","MTA2","MTA3","ARID1A","ARID2","PHC3","PHC1","PHC2","BMI1","YY1","JARID2","SUZ12","EED","HNF1A","NCOR1","ARID4A","MEN1","RBBP5","WDR5","CHAF1A","CHAF1B","TDG","DAXX","SIN3A","SIN3B","ARID1B","ACTL6B"),
  helicase=c("ATRX","HELLS","INO80","SMARCA1","SMARCA2","SMARCA4","SMARCA5","SMARCB1","SMARCC1","SMARCC2","SMARCD1","SMARCD2","SMARCD3","SMARCE1","CHD1","CHD1L","CHD2","CHD3","CHD4","CHD5","CHD6","CHD7","CHD8","CHD9"),
  others=c("PADI1","PADI2","PADI3","PADI4","PADI6","UBE2A","UBE2B","UBE2E1","UBE2I","UBR7","USP22","USP27X","USP51","BAP1")
)
```

##Expression changes
We have analyzed significant differences in mRNA expression of ERGs across different cancer types comparing TCGA tumor samples and TCGA adjacent tissus + GTEX database normal tissus.

We cannot directly compare data from this two databases together. There is diffences in sample and data processing. 
Hopfully this have been solved in Quingguo et al. 2018. for 18 cancer type :

<cite>Unifying cancer and normal RNA sequencing data from different sources
Qingguo Wang, Joshua Armenia, Chao Zhang, Alexander V. Penson, Ed Reznik, Liguo Zhang, Thais Minet, Angelica Ochoa, Benjamin E. Gross, Christine A. Iacobuzio-Donahue, Doron Betel, Barry S. Taylor, Jianjiong Gao & Nikolaus Schultz
Sci Data. 2018 Apr 17;5:180061. doi: 10.1038/sdata.2018.61.</cite>

```{r include = TRUE}
load("../expression.rdata")
```

###Fig3A Gene Expression (TPM) in Tumor tissus from TCGA (red) vs adjacent tissu from TCGA + normal tissus from GTEx (blue).

* There is only adjacent tissus for CHOL and HNSC

```{r include = TRUE}
#Calculate log2FC using mean exp of normal samples. log2FC(Ti)= log2(exp(Ti)/mean(exp(Normaux))
logex<-expression[ group=="normal",  .(mean=mean(exp)), by=c("cancer","genes")]
logex<-merge(expression,logex,by=c("cancer","genes"))
logex[ , logFC:=log2(exp/mean), ]
logex[ , FC:=exp/mean, ]

tmp<-logex[ , .( log_mean_exp=mean(log2(exp))), by=c("cancer","genes","group") ]

fig3A<-ggplot( tmp, aes(x=cancer, y=log_mean_exp) ) + 
  geom_boxplot(aes(fill=group), outlier.shape=NA) + 
  scale_y_continuous(limits=c(6,15)) +
  scale_fill_manual(values=palette1) +
  theme(axis.text.x=element_text(angle=50,hjust=1, size=30), axis.title.x=element_blank() ) +
  theme(axis.text.y=element_text(size=12), axis.title.y=element_text(size=20)) +
  theme(legend.text=element_text(size=30), legend.title=element_text(size=30)  ) +
  ylab("mean Log2(TPM)")
```

```{r echo = FALSE, fig.height = 10, fig.width = 20, fig.align = "center"}
plot(fig3A)
```

```{r include = FALSE}
jpeg("figure3A.jpeg", width=2800, height=1200, res = 200)
plot(fig3A)
dev.off()
```

###Figure 3B RNA EXPRESSION (log2FC) FILTERED FOR CONSISTENCY > 85% AND PENETRANCE > 15%

We plotting RNA expression (TPM) of ERG filtered for consistency and penetrance of deregulation .

* CONSISTENCY = 95% of sample change in the same direction
* PENETRANCE =  Fold Change is superior to 15%

```{r include = TRUE}
#calculate consistency and penetrance
logex<-logex[group=="tumor",]
logex<-logex[!is.na(logFC) , .( .N,  N2=sum(abs(logFC)>1), up=sum(logFC>1), down=sum(logFC < -1) ), by=c("cancer","genes")]
logex$pUp<-logex$up/logex$N2
logex$pDown<-logex$down/logex$N2
logex$pChange<-logex$N2/logex$N
logex$pC<-pmax(logex$pUp,logex$pDown)
logex$pC[logex$pC>0.95]<-1
logex$pC[logex$pC>0.90 & logex$pC<=0.95 ]<-0.95
logex$pC[logex$pC>0.85 & logex$pC<=0.90 ]<-0.90
logex$group<-"up"
logex[pUp<pDown]$group<-"down"

#filter and format data
tmp<-logex[ ( pUp > 0.85 | pDown > 0.85 ) & pChange >0.15 , .( .N, genes), by=c("cancer","pC","group")]
tmp$pC<-as.factor(tmp$pC*100)
tmp$group<-as.factor(tmp$group)
order<-c("ESCA","STAD","LIHC","READ","CHOL","THCA","CESC","COAD","BRCA","LUSC","PRAD","BLCA","UCEC","LUAD","KICH","HNSC","KIRC","KIRP")
tmp$cancer<-factor(tmp$cancer,levels=order)

#build figure
fig3B<-ggplot( data=tmp, aes(x=cancer, y=pC, group=group )) + geom_point(aes(size=N, color=group) ) + 
  theme_minimal() + 
  scale_color_manual(values=c(amp_col,del_col)) + ggtitle("Consistency in the direction of expression accross samples (%)") +     
  theme(axis.title=element_text(size=6,face="bold")) +
  theme(axis.text.x=element_text(angle=50,hjust=1, size=12), axis.title.x=element_blank()) +
  theme(axis.text.y=element_text(size=12), axis.title.y=element_blank() )
```

```{r echo = FALSE, fig.height = 8, fig.width = 16, fig.align = "center"}
plot(fig3B)
```

```{r include = FALSE}
jpeg("figure3B.jpeg", width=1200, height=600, res = 200)
plot(fig3B)
dev.off()
```

## Figure 3C : Clustering of cancers using genes with logFC > 2

```{r echo = TRUE, fig.height = 17, fig.width = 17, fig.align = "center"}
meanexp<-expression[ group=="normal",  .(mean=mean(exp)), by=c("cancer","genes")]
meanexp<-merge(expression,meanexp,by=c("cancer","genes"))
meanexp[ , logFC:=log2(exp/mean), ]
meanexp[ , FC:=exp/mean, ]

foo<-acast(meanexp[ abs(logFC)>2, ], genes~cancer, value.var="logFC",fun.aggregate=mean)
foo[!is.finite(foo)]<-NA
foo<-foo[complete.cases(foo),]
heatmap.2(foo,trace="none", col=palette, cexRow = 1, cexCol = 0.8, keysize=1, key.title="")
```

## Differential expression analyses with EdgeR

We run a differential expression analyses with edgeR package and draw volcano plots

```{r include = TRUE, eval = params$reload}
cancers<-unique(expression$cancer)
final<-data.table()
for (cancer_ in cancers){
  
  pdata<-unique(expression[ cancer==cancer_, c("samples","database","group") ])
  data<-acast(expression[cancer==cancer_,], genes~samples+database+group, value.var="exp" )
  
  #design
  design<-model.matrix(~0 + group, data=pdata)
  
  #remove genes that are unexpressed or very lowly expressed in the samples
  cpm_log <- cpm(data, log = TRUE)
  median_log2_cpm <- apply(cpm_log, 1, median)
  #hist(median_log2_cpm)
  expr_cutoff <- -1
  #abline(v = expr_cutoff, col = "red", lwd = 3)
  sum(median_log2_cpm > expr_cutoff)
  data_clean <- data[median_log2_cpm > expr_cutoff, ]
  
  #log2 cpm
  cpm_log <- cpm(data_clean, log = TRUE)
  
  #differential expression analysis
  y <- DGEList(counts = data_clean, group = pdata$group)
  y <- calcNormFactors(y)
  y <- estimateDisp(y)
  et <- exactTest(y)
  results_edgeR <- topTags(et, n = nrow(data_clean), sort.by = "none")
  head(results_edgeR$table)
  table<-results_edgeR$table[rownames(results_edgeR$table) %in% genes426,]
  table$genes<-rownames(table)
  table<-as.data.table(table)
    
  #save data
  table$cancer<-cancer_
  final<-rbind(final, table,fill=T)
}
save(final,file="final.rdata")
```

```{r include = TRUE, eval = !params$reload}
load("../final.rdata")
```

## Differential expression analyses : Volcano plots

```{r include = TRUE}
HM<-c(genelist$HM_e, genelist$HM_r,genelist$HM_w)
HA<-c(genelist$HA_e, genelist$HA_r, genelist$HA_w)
DM<-c(genelist$DM_e, genelist$DM_r, genelist$DM_w)

#build figure
fig3C<-ggplot()+
  geom_point( data=final, color="#777777", aes(x=logFC,y=-log10(PValue)) ) +
  #geom_point( data=final[FDR < 0.05,], color="blue", aes(x=logFC,y=-log10(PValue)) ) +
  geom_point( data=final[genes %in% HM,], color="#00ff00", alpha=0.5, aes(x=logFC,y=-log10(PValue)) ) +
  geom_point( data=final[genes %in% HA,], color="#ff0000", alpha=0.5, aes(x=logFC,y=-log10(PValue)) ) +
  geom_point( data=final[genes %in% DM,], color="#0000ff", alpha=0.5, aes(x=logFC,y=-log10(PValue)) ) +
  #geom_text_repel( data=final[FDR < 0.05 & abs(logFC) > 2 & genes %in% unlist(genelist)], aes(x=logFC,y=-log10(PValue), label=genes) ,angle = 45 ,size=2 ) +
  facet_wrap(~cancer, ncol=6, scales = "free")
```

* HM (green)
* HA (red)
* DM (blue)

```{r echo = FALSE, fig.height = 10, fig.width = 17, fig.align = "center"}
plot(fig3C)
```

```{r include = FALSE}
jpeg("figure3C.jpeg", width=2800, height=1200, res = 200)
plot(fig3C)
dev.off()
```

## Volcano plots cancer by cancer

```{r include = TRUE}
cancers<-unique(expression$cancer)
for (cancer_ in cancers){
  
  foo<-final[cancer==cancer_]
  sel<-head(foo[order(FDR)])$genes
  fig<-ggplot()+ ggtitle(cancer_) +
      geom_point( data=foo, color="black", aes(x=logFC,y=-log10(PValue)) ) +
      geom_point( data=foo[FDR < 0.05,], color="blue", aes(x=logFC,y=-log10(PValue)) ) +
      geom_point( data=foo[FDR < 0.05 & abs(logFC) > 1,], color="red", aes(x=logFC,y=-log10(PValue)) ) +
      geom_text_repel( data=foo[FDR < 0.05 & abs(logFC) > 2,], aes(x=logFC,y=-log10(PValue), label=genes ))
  plot(fig)
  
  jpeg(paste0("Volcano_plots/",cancer_,"_volcano.jpeg"), width=2000, height=2000, res = 200)
  plot(fig)
  dev.off()
}
```

## Circle plot showing correlation between Copy Number Variation (CNV) and Expression (TPM)

Correlation between variation of CNV and expression is represented by Circos plot. They were drawn with the Circlize package. 
A positive correlation reflect an amplification of copy number associated with upregulation of the gene. Or a deletion associated with downregulation.
Contrary negative correlation reflect a deletion of copy number associated with upregulation of the gene. Or amplification associated with downregulation.

```{r show=TRUE}
Circle<-function(){

	#remove not significant
	dtCor$mCor[dtCor$qvalue>0.05]<-NA

	#genes selection
	sel <- dtCor[ , .( maxCorr=max(abs(mCor), na.rm=T), sumNA=sum(is.na(mCor)), maxCNV=max(mCNV), minCNV=min(mCNV)  ), by=genes ]
	sel$maxCorr[!is.finite(sel$maxCorr)]<-0
	selection<-sel[ which(sel$maxCorr>0.3 & sel$sumNA<32), genes]

	#create matrix of correlation
	foo<-acast(dtCor[dtCor$genes %in% selection], genes~cancer, value.var="mCor")
	foo[is.na(foo)]<-(0)

	#get genomic positions
	k = keys(Homo.sapiens, keytype="SYMBOL")
	res <- select(Homo.sapiens, keys = k, columns =c("TXCHROM","TXSTART","TXEND"), keytype="SYMBOL", multiVals='first')
	kermit<-merge(foo, res[which(!duplicated(res)), ], by.x="row.names", by.y="SYMBOL")
	c=ncol(kermit)
	bed<-kermit[,c((c-2):c,1:(c-3))]
	colnames(bed)[1:4]<-c("chr","start","end","symbol")
	bed<-bed[!duplicated(bed$symbol),]
	bed<-bed[!is.na(bed$chr),]
	return(bed)

}
```

```{r show=TRUE}
drawFullCircle<-function(bed,pal){

	circos.clear()
	circos.par()
	circos.initializeWithIdeogram()
	circos.genomicPosTransformLines(bed, posTransform = posTransform.default, horizontalLine = "top")
	om = circos.par("track.margin")
	oc = circos.par("cell.padding")
	circos.par(track.margin = c(om[1], 0), cell.padding = c(0, 0, 0, 0))
	f = colorRamp2(breaks = c(-1, -0.5, 0, 0.5, 1), colors = pal)
	circos.genomicTrackPlotRegion(bed, stack = TRUE, panel.fun = function(region, value, ...) {
	  
	  circos.genomicRect(region, value, col = f(value[[1]]), 
		             border = f(value[[1]]), lwd = 0.1, posTransform = posTransform.default, ...)
	  
	}, bg.border = NA, track.height = 0.5)
	circos.par(track.margin = om, cell.padding = oc)

}
```

```{r include = TRUE}
load("../data.rdata")
```

```{r show=TRUE}
# CNV vs EXP amplifications only
suppressWarnings(dtCor <- data[mutStatus==0 & !is.na(exp) & cnv>=0,  .(mCor = cor.test(exp,cnv, use = "na.or.complete" )$estimate, .N, t=cor.test(exp,cnv, use = "na.or.complete" )$statistic, pvalue=cor.test(exp,cnv, use = "na.or.complete" )$p.value, mCNV=mean(cnv)), by=c("cancer","genes")])
dtCor$qvalue <- p.adjust(dtCor$pvalue, method="fdr")
suppressWarnings(bed1<-Circle())
```

```{r show=TRUE}
# CNV vs EXP deletion only
suppressWarnings(dtCor <- data[mutStatus==0 & !is.na(exp) & cnv<=0,  .(mCor = cor.test(exp,cnv, use = "na.or.complete" )$estimate, .N, t=cor.test(exp,cnv, use = "na.or.complete" )$statistic, pvalue=cor.test(exp,cnv, use = "na.or.complete" )$p.value, mCNV=mean(cnv)), by=c("cancer","genes")])
dtCor$qvalue <- p.adjust(dtCor$pvalue, method="fdr")
suppressWarnings(bed2<-Circle())
```

```{r echo=FALSE, fig.height = 7, fig.width = 17, fig.align = "center"}
par(mfrow = c(1,3))
drawFullCircle(bed1,pal1)
title("Amplified", cex.main=4, line = -2)
drawFullCircle(bed2,pal2)
title("Deleted", cex.main=4, line = -2)
```

```{r include = FALSE}
jpeg("figure3f.jpeg", width=2800, height=1200, res = 200)
drawFullCircle(bed1,pal1)
dev.off()

jpeg("figure3g.jpeg", width=2800, height=1200, res = 200)
drawFullCircle(bed2,pal2)
dev.off()
```

###Fig3H Mutual exclusivity

To obtain complete overview across different cancer types we plotted all three deregulations together, mutational status, CNV alteration and changes in expression.

* Green triangles are deleted genes, red one are amplified while blue circle are only mutated.
* Expression (Zscore) is reported on the Y axis. We calculated the mean by genes over samples.
* We calculated mean of cnv+3 to avoid issues with negative values.

```{r include = TRUE}
#parameters
snvcutoff= 0.01   #threshold for mutation = 1%
cnvcutoff= 0.01   #threshold for cnv = 1%
```

```{r include = TRUE}
#select and filter and format data
mutex<-data[, .(meanZscore=mean(zscore,na.rm=T), meanMut=mean(mutStatus), meanCNV=mean(cnv+3) ), by=c("cancer","genes","geneclass")]
mutex$mutStatus<-(mutex$meanMut>snvcutoff)
mutex$cnv<-mutex$meanCNV
mutex[meanCNV>=(3+cnvcutoff)]$cnv<-1
mutex[meanCNV<(3+cnvcutoff)]$cnv<-0
mutex[meanCNV<=(3-cnvcutoff)]$cnv<--1
mutex$cnv<-factor(mutex$cnv)
mutex$mutStatus<-factor(mutex$mutStatus)
levels(mutex$mutStatus)<-c("normal","mutated")
mutex$group<-factor(paste0(mutex$mutStatus, mutex$cnv), levels=c("normal-1", "normal0","normal1","mutated-1", "mutated0","mutated1") )
mutex<-mutex[ !(mutStatus==0 & cnv==0), ]
mutex$cnv<-factor(mutex$cnv,levels=c("-1","1","0"))

#order cancers for better display
group1<-c("UCEC","ACC","LUAD","BLCA","CESC","UCS","MESO","KICH","CHOL","COAD/READ","LUSC","BRCA","GBM","SKCM","UVM","DLBC")
group2<-c("TGCT","STAD","OV","ESCA","HNSC","LIHC","PAAD","KIRC","SARC","KIRP","PCPG","PRAD","LGG","THYM","LAML","THCA")
mutex$cancer<-factor(mutex$cancer,levels=c(group1,group2))

#Build Figure
p0<-ggplot(data=mutex[cancer %in% group2, ], aes(x=cancer, y=meanZscore, fill=cnv))  + 
  geom_point(position= position_jitterdodge(), aes(color=cnv, shape=cnv)) + 
  scale_shape_manual(values=c(25,24,21), labels=c("deletion","amplification","normal"),name="CNV" ) + 
  theme(axis.text.x=element_blank(), axis.ticks.x=element_blank(), legend.text=element_text(size=12), panel.spacing.x=unit(0.1,"lines")) + 
  geom_hline(yintercept=0,color="black") + 
  scale_y_continuous(limits = c(-2, 2.5)) +
  guides(colour=FALSE, shape=FALSE, fill=FALSE) +
  scale_colour_manual(values=palette4) +
  scale_fill_manual( values=palette5) +
  theme(panel.grid.major = element_blank(), axis.title.y=element_blank(), panel.grid.minor = element_blank(), plot.margin=unit(c(0,1,0,1), "cm") ) +
  facet_wrap(~cancer, scales="free_x", nrow=1)

p1<-ggplot(data=mutex[cancer %in% group1, ], aes(x=cancer, y=meanZscore, fill=cnv))  + 
  geom_point(position= position_jitterdodge(), aes(color=cnv, shape=cnv)) + 
  scale_shape_manual(values=c(25,24,21), labels=c("deletion","amplification","normal"),name="CNV" ) + 
  theme(axis.text.x=element_blank(), axis.ticks.x=element_blank(), legend.text=element_text(size=12), panel.spacing.x=unit(0.1,"lines")) + 
  geom_hline(yintercept=0,color="black") + 
  scale_y_continuous(limits = c(-2, 2)) +
  guides(colour=FALSE, shape=FALSE, fill=FALSE) +
  scale_colour_manual(values=palette4) +
  scale_fill_manual( values=palette5) +
  theme(axis.title.x=element_blank(), axis.title.y=element_blank(), panel.grid.major = element_blank(), panel.grid.minor = element_blank(), strip.text=element_blank(), plot.margin=unit(c(0,1,0,1), "cm") )  +
  facet_wrap(~cancer, scales="free_x", nrow=1)

p2<-ggplot(data=mutex[cancer %in% group1, ], aes(x=cancer, y=meanZscore, fill=cnv)) + 
  geom_point(position= position_jitterdodge(), aes(color=cnv, shape=cnv)) + 
  scale_shape_manual(values=c(25,24,21), labels=c("deletion","amplification","normal"),name="CNV" ) + 
  theme(axis.title.x=element_blank(), axis.text.x=element_blank(),axis.ticks.x=element_blank(), panel.spacing.x=unit(0.1,"lines")) + 
  scale_y_continuous(limits = c(2, 25)) +
  guides(colour=FALSE, shape=FALSE, fill=FALSE) +
  scale_colour_manual(values=palette4) +
  scale_fill_manual( values=palette5) +
  theme(panel.grid.major = element_blank(), axis.title.y=element_blank(), panel.grid.minor = element_blank(), strip.text=element_blank(), plot.margin=unit(c(0,1,0,1), "cm") ) +
  facet_wrap(~cancer, scales="free_x", nrow=1)

p3<-ggplot(data=mutex[cancer %in% group1, ], aes(x=cancer, y=meanZscore, fill=cnv)) + 
  geom_point(position= position_jitterdodge(), aes(color=cnv, shape=cnv)) + 
  scale_shape_manual(values=c(25,24,21), labels=c("deletion","amplification","normal"),name="CNV" ) + 
  theme(axis.title.x=element_blank(), axis.text.x=element_blank(),axis.ticks.x=element_blank(), panel.spacing.x=unit(0.1,"lines")) +
  scale_y_continuous(limits = c(80, 100), breaks=seq(80,100,20)) +
  guides(colour=FALSE, shape=FALSE, fill=FALSE) +
  scale_colour_manual(values=palette4) +
  scale_fill_manual( values=palette5) +
  theme(panel.grid.major = element_blank(), axis.title.y=element_blank(), panel.grid.minor = element_blank(), plot.margin=unit(c(1,1,0,1), "cm") ) +
  facet_wrap(~cancer, scales="free_x", nrow=1)

gZ <- ggplotGrob(p0)
gA <- ggplotGrob(p1)
gB <- ggplotGrob(p2)
gC <- ggplotGrob(p3)
maxWidth = grid::unit.pmax(gZ$widths[2:5], gA$widths[2:5], gB$widths[2:5], gC$widths[2:5] )
gZ$widths[2:5] <- as.list(maxWidth)
gA$widths[2:5] <- as.list(maxWidth)
gB$widths[2:5] <- as.list(maxWidth)
gC$widths[2:5] <- as.list(maxWidth)
```

```{r echo = FALSE, fig.height = 17, fig.width = 17, fig.align = "center"}
grid.arrange(gC, gB, gA, gZ, ncol=1, heights=c(2,2,6,6))
```

```{r include = FALSE}
jpeg("figure3h.jpeg", width=2600, height=1800, res = 200)
grid.arrange(gC, gB, gA, gZ, ncol=1, heights=c(2,2,6,6))
dev.off()
```
